OvmfPkg/PlatformPei/PlatformId.c

```diff
@@ -0,0 +1,660 @@
+ * 0 - Raw data in the locator itself.
+ * 1 - URI in rtf2396 format.
+ * 2 - local device path in EFI_DEVICE_PATH_PROTOCOL format.
+ * 3 - UEFI variable (EFI_GUID, 00-terminated UCS2 string)
+ *
+ * Each of these could potentially be validated before registering, but they
+ * aren't necessarily loaded in the PEI phase this runs in.
+**/
+#define LOCATOR_TYPE_RAW_DATA       0
+#define LOCATOR_TYPE_URI            1
+#define LOCATOR_TYPE_DEVICE_PATH    2
+#define LOCATOR_TYPE_UEFI_VARIABLE  3
+
+/**
+ * If Buffer + Offset does not exceed BufferEnd, then sets *Result to
+ * Buffer + Count.
+ *
+ * @param[in]  Buffer  The pointer to offset.
+ * @param[in]  BufferEnd  The end of the buffer.
+ * @param[in]  Offset  The number of bytes to offset Buffer.
+ * @param[out] Result  Buffer + Count
+ *
+ * @retval EFI_SUCCESS  The read was successful.
+ * @retval EFI_BUFFER_TOO_SMALL  The read would exceed BufferEnd.
+*/
+STATIC
+EFI_STATUS
+SafeOffset (
+  IN  UINT8  *Buffer,
+  IN  UINT8  *BufferEnd,
+  IN  UINTN  Offset,
+  OUT UINT8  **Result
+  )
+{
+  if ((Buffer > BufferEnd) || ((UINTN)(BufferEnd - Buffer) < Offset)) {
+    return EFI_BUFFER_TOO_SMALL;
+  }
+
+  *Result = Buffer + Offset;
+  return EFI_SUCCESS;
+}
+
+/**
+ * Reads a 32-bit integer from Buffer and returns it in *Value, and the pointer
+ * after the integer in Buffer in *End.
+ *
+ * @param[in] Buffer  The buffer to read from.
+ * @param[in] BufferEnd  The end of the buffer.
+ * @param[out] Value  The value read.
+ * @param[out] End  The byte after the value.
+ *
+ * @retval EFI_SUCCESS  The value was read successfully.
+ * @retval EFI_BUFFER_TOO_SMALL  The buffer is smaller than 32 bits.
+*/
+STATIC
+EFI_STATUS
+ParseUint32 (
+  IN  UINT8   *Buffer,
+  IN  UINT8   *BufferEnd,
+  OUT UINT32  *Value,
+  OUT UINT8   **End
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = SafeOffset (Buffer, BufferEnd, sizeof (UINT32), End);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  *Value = *(UINT32 *)Buffer;
+  return EFI_SUCCESS;
+}
+
+/**
+ * Reads a first byte as a size, then confirms that Buffer + 1 + size
+ * does not exceed BufferEnd and that the string is NUL-terminated.
+ *
+ * @param[in] Buffer  The buffer to read from.
+ * @param[in] BufferEnd  The end of the buffer.
+ * @param[out] Start  The start of the string.
+ * @param[out] End  The byte after the NUL-terminator.
+ *
+ * @retval EFI_SUCCESS  The string was read successfully.
+ * @retval EFI_BUFFER_TOO_SMALL  The string's declared size was too large for
+ *                               the buffer.
+ * @retval EFI_INVALID_PARAMETER  The string was not NUL-terminated.
+ */
+STATIC
+EFI_STATUS
+ConsumeAsciiByteStr (
+  IN  UINT8  *Buffer,
+  IN  UINT8  *BufferEnd,
+  OUT UINT8  **Start,
+  OUT UINT8  **End
+  )
+{
+  EFI_STATUS  Status;
+  UINT8       EvtSize;
+
+  Status = SafeOffset (Buffer, BufferEnd, sizeof (EvtSize), Start);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  EvtSize = *Buffer;

```

Comment:
To follow your ParseUint32() logic, this seems ParseUint8().

Why not implement a ParseUint8() function?

Reply:
Removed for uninterpreted single file solution.